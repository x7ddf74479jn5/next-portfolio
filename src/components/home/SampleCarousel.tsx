import Image from "next/image";
import * as React from "react";
import styles from "src/styles/components/home/SampleCarousel.module.scss";
import SwiperCore, { Autoplay, Navigation, Pagination, Scrollbar } from "swiper";
import { Swiper, SwiperSlide } from "swiper/react";

SwiperCore.use([Navigation, Pagination, Scrollbar, Autoplay]);

import Link from "next/link";
import { useEffect } from "react";
import type { Sample } from "src/contents/samples";

type SwiperProps = {
  samples: Sample[];
};

const SampleCarousel: React.FC<SwiperProps> = ({ samples }) => {
  const [mounted, setMounted] = React.useState<boolean>(false);

  useEffect(() => {
    setMounted(true);
  }, []);

  return mounted ? (
    <Swiper
      pagination={{ clickable: true }}
      navigation
      loop={true}
      spaceBetween={0}
      speed={1000}
      centeredSlides={true}
      autoplay={{ delay: 3000 }}
      slidesPerView={1}
      breakpoints={{
        1024: {
          slidesPerView: 3,
          spaceBetween: 10,
          autoplay: {
            delay: 5000,
          },
        },
      }}
    >
      {samples.map((sample, index) => {
        return (
          <SwiperSlide key={`swiper-${index}`}>
            <article className={styles.gridListItem}>
              <Link href={sample.link}>
                <a className={styles.textLineNone} target="_self">
                  <h3>{sample.title}</h3>
                  <div className={styles.mediaThumb}>
                    <Image loading="eager" layout="fill" src={sample.imgSrc} alt={sample.imgAlt} objectFit="contain" />
                  </div>
                  <p>{sample.description}</p>
                </a>
              </Link>
            </article>
          </SwiperSlide>
        );
      })}
    </Swiper>
  ) : null;
};

export default SampleCarousel;
