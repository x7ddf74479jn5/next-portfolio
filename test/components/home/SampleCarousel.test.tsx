import SampleCarousel from "src/components/home/SampleCarousel";
import type { RenderResult } from "test/test-utils";
import { reTestCase } from "test/test-utils";
import { render } from "test/test-utils";

import type { Sample } from "../../../src/contents/samples";

jest.mock("next/image", () => {
  const Image = ({ src, alt }: { src: string; alt: string }): JSX.Element => {
    return <img src={src} alt={alt} />;
  };
  return Image;
});

describe("SampleCarousel", () => {
  let renderResult: RenderResult;
  const sample: Sample[] = [
    {
      id: "id",
      imgSrc: "src.png",
      imgAlt: "alt",
      category: "category",
      link: "link",
      title: "title",
      description: "description",
    },
  ];

  beforeEach(() => {
    renderResult = render(<SampleCarousel samples={sample} />);
  });

  it("matches snapshot", () => {
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it("rendered correctly", () => {
    // duplicate + original + duplicate
    const articleElements = renderResult.getAllByRole("article");
    expect(articleElements).toHaveLength(3);

    const anchorElements = renderResult.getAllByRole("link");
    expect(anchorElements[0]).toHaveAttribute("href", reTestCase.anyWord);

    const headingElements = renderResult.getAllByRole("heading");
    expect(headingElements[0]).toHaveTextContent("title");

    const imgElements = renderResult.getAllByRole("img");
    expect(imgElements[0]).toHaveAttribute("src", reTestCase.anyImage);
    expect(imgElements[0]).toHaveAttribute("alt", reTestCase.anyWord);
  });
});
