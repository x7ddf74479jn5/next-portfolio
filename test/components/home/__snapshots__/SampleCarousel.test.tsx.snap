// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleCarousel matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="swiper-container swiper-container-initialized swiper-container-horizontal"
  >
    <div
      class="swiper-button-prev"
    />
    <div
      class="swiper-button-next"
    />
    <div
      class="swiper-pagination swiper-pagination-clickable swiper-pagination-bullets"
    />
    <div
      class="swiper-wrapper"
      style="transition-duration: 0ms; transform: translate3d(NaNpx, 0px, 0px);"
    >
      <div
        class="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index="0"
      >
        <article
          class="gridListItem"
        >
          <a
            class="textLineNone"
            href="/link"
            target="_self"
          >
            <h3>
              title
            </h3>
            <div
              class="mediaThumb"
            >
              <img
                alt="alt"
                src="src.png"
              />
            </div>
            <p>
              description
            </p>
          </a>
        </article>
      </div>
      <div
        class="swiper-slide"
        data-swiper-slide-index="0"
      >
        <article
          class="gridListItem"
        >
          <a
            class="textLineNone"
            href="/link"
            target="_self"
          >
            <h3>
              title
            </h3>
            <div
              class="mediaThumb"
            >
              <img
                alt="alt"
                src="src.png"
              />
            </div>
            <p>
              description
            </p>
          </a>
        </article>
      </div>
      <div
        class="swiper-slide swiper-slide-duplicate"
        data-swiper-slide-index="0"
      >
        <article
          class="gridListItem"
        >
          <a
            class="textLineNone"
            href="/link"
            target="_self"
          >
            <h3>
              title
            </h3>
            <div
              class="mediaThumb"
            >
              <img
                alt="alt"
                src="src.png"
              />
            </div>
            <p>
              description
            </p>
          </a>
        </article>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
